Update the resource revisions guide using nested collections. (#1118)
googleapis/aip: aip/general/0162.md

APIs **may** store a revision history for a resource if it is useful to users.

> APIs implementing resources with a revision history **should** abstract resource revisions as nested collection of the resource. The

Reviewer: There should be an explanation of what situations revisions should be used vs not used. Can that be added?

Reviewer: Sorry I guess a better question: why should an API who wants a revision history *not* follow this pattern? The wording here implies says basically: "there are valid cases where you would implement revisions, but not do it in the way this AIP states". What would those cases be?

> resource revision type **should** inherit from the parent resource and has a field being the parent resource type.

Reviewer: Id make this a must. If you're choosing to implement the pattern, there's not a lot of value in only doing part of it.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;

   // The snapshot of the book
   Book book = 2 [(google.api.field_behavior) = IMMUTABLE];

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // tags set on the resource revisions by service or user.
  repeated string tags = 4;
  [(google.api.field_behavior) = OUTPUT_ONLY];
}
```

- The resource revision **must** contain a field representing the snapshot of the parent resource.
- The resource revision **must** contain a `create_time` field, which

>   **should** be a `google.protobuf.Timestamp` (see [AIP-142][]).

Reviewer: let's make this a must?

Author: updated

- The resource revision **may** contain a repeated field `tags` to represent tags that was set on the resource revision.

### Referencing revisions

> When referring to specific revision of a resource, APIs must use the following syntax

Reviewer: I'd rewrite this section is "resource names for revisions".

There's already guidance on how to refer to other resources: https://google.aip.dev/122#fields-representing-another-resource. It's using the resource name.

So this should just suggest a resource name (`/foo/{}/revisions/{}`).

Reviewer 2: I agree - I don't think this should be written in terms of "the following syntax". It should be "revisions ***must*** exist in a subcollection of the original resource, named `revisions`" and then the example.

`{resource_name}/revisions/{revision_id}`. For example:

publishers/123/books/les-miserables/revisions/c7cfa2a8

> ### Getting a revision

Reviewer: This section isn't really needed anymore, since a revision is no different from a child collection.

I think this AIP should talk about how revisions should be modelled as as a sub-collection of a parent resource instead. WDYT?

Reviewer: All the guidance in this section is redundant with https://google.aip.dev/131

Author: SG. Still kept this section and mention that this is just standard Get method. Should I remove this section?

Reviewer: Yeah I'd just get rid of it. Or maybe just add it in the top-level guidance: it's just an 1-liner.

APIs implementing resource revisions in the standard `Get` method ([AIP-131][]):

```proto
// Gets the details of a book revision.
rpc GetBookRevision(GetBookRevisionRequest) returns (BookRevision) {
  option (google.api.http) = {
    get: "/v1/{name=publishers/*/books/*/revisions/*}"
  };
}

message GetBookRevisionRequest {
  // The name of the book revision.
  //   Example: publishers/123/books/les-miserables
  //
  // In order to retrieve a previous revision of the book, also provide
  // the revision ID.
  //   Example: publishers/123/books/les-miserables/revisions/c7cfa2a8
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];
}
```

If the user passes a revision ID that does not exist, the API must fail with a `NOT_FOUND` error.

The returned resource revision `should` be a different resource type from the parent resource type.

### Tagging revisions

> There are two types of tags: Server defined and user specified.

Reviewer: can these two be defined? As in:

```
There are two types of tags: service defined and user-specified

services **may** reserve specific tags to be server-defined (e.g. `current`), which are read-only and managed by the service.

user-specified tags are...
```

> Service may support `current`, `latest`, `oldest`, etc as preserved keyword tags.

Reviewer: This should be written in AIP style:

```
- Revision IDs *may* be reserved to imply a specific meaning (e.g. `latest`).
```

Reviewer: and also s/preserved/reserved/g.

> GET /v1/publishers/{publisher}/books/{book}/revisions/{preserved_keyword}

Reviewer: nit: quote.

> Note that `publishers/{publisher}/books/{book}/revisions/latest` and `publishers/{publisher}/books/{book}` can differ. The latest revision may differ from the current status of the resource.

Reviewer: Is something special about `latest` as a revision? Above, it's mentioned as a possible reserved keyword, but there's no explanation as to what `latest` should mean.

I feel like we should be adding guidance here.

Author: added some explanation for latest.

APIs implementing resource revisions **may** provide a mechanism for users to
tag a specific revision with a user provided name by implementing a "Tag
Revision" custom method:

```proto
rpc TagBookRevision(TagBookRevisionRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*/revisions/*}:tagRevision"
    body: "*"
  };
}
```

```proto
message TagBookRevisionRequest {
  // The name of the book revision to be tagged.
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];

  // The tag to apply.
  // The tag should be at most 40 characters, and match `[a-z][a-z0-9-]{3,38}[a-z0-9]`.
  string tag = 2 [(google.api.field_behavior) = REQUIRED];
}
```

> - The `name` field **should** require an explicit resource revision to be provided.

Reviewer: I can't comment on undiffed lines, so:

Note that the following is still in the AIP:

```
Once a revision is tagged, the API **must** support using the tag in place of
  the revision ID in `name` fields.
```

so you can do things like `revisions/current` as well as `revisions/{revision_id}`.

Is that desirable? It's basically creating a second resource id in that case. I kind of feel like at that point, the `tag` API just creates a new resource.

Author: The tag is more like alias. I am thinking that it doesn't create new resource. Yes, it is a second resource id. revisions/abcd and revisions/myfavorite can point to the same revision resource.

When the revision resource like abcd is deleted, all the tags set by the user and assoicated with abc should be released.

Reviewer: I think that's ok? although it does create weird conflicts for ownership.

regardless that provision has been in there so let's just keep it.

APIs **may** store a revision history for a resource if it is useful to users.

APIs implementing resources with a revision history **should** abstract resource
revisions as nested collection of the resource. Sometimes, the revisions 
collection can be a top level collection, exception includes:

- If resource revisions are meant to have longer lifespan than the parent 
resource. In other words, resource revisions exist after the resource deletion.

The resource revision type
**must** inherit from the parent resource and has a field being the parent
resource type.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;

   // The snapshot of the book
   Book book = 2 [(google.api.field_behavior) = IMMUTABLE];

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // tags set on the resource revisions by service or user.
  repeated string tags = 4;
  [(google.api.field_behavior) = OUTPUT_ONLY];
}
```

> - The resource revision **must** contain a field representing the snapshot of

Reviewer: ```suggestion
- The resource revision **must** contain a field with a message type of the parent resource, with a field name of the resource singular.
```

Rather than introduce the term "snapshot", I think it's easier to be precise.

  the parent resource.
- The resource revision **must** contain a `create_time` field, which
  **must** be a `google.protobuf.Timestamp` (see [AIP-142][]).
- The resource revision **may** contain a repeated field `tags` to represent
  tags that was set on the resource revision.

> ### When resource revisions are useful

Reviewer: this should probably be at the top, under "guidance". The first part of a doc should probably explain why this is a relevant AIP.

- When it is valuable to expose older versions of a resource via an API. This
  can avoid the overhead of the customers having to write their own API to store
  and enable retrieval of revisions.
- Other resources depend on different revisions of a resource.
- There is a need to visualize the change of a resource over time.

### Resource names for revisions

Revisions **must** exist in a subcollection of the original resource. When 
referring to specific revision of a resource, the subcollection name **must**

> named `revisions`. Resource revision has the name with the format

Reviewer: ```suggestion
be named `revisions`. Resource revisions have names with the format
```

 `{resource_name}/revisions/{revision_id}`. For example:

publishers/123/books/les-miserables/revisions/c7cfa2a8

### Creating revisions

Depending on the resource, different APIs may have different strategies for
when to create a new revision, such as:

- Create a new revision any time that there is a change
- Create a new revision when something important happens
- Create a new revision when the user specifically asks

APIs **may** use any of these strategies. APIs that want to create a revision on

> user request should handle this with a custom method on the resource:

Reviewer: ```suggestion
user request **must** handle this with a custom method on the resource:
```

Any reason to not make this a must? if it is a should, what are the exceptions?

```proto
rpc CommitBookRevision(CommitBookRevisionRequest) returns (BookRevision) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*}:commitRevision"
    body: "*"
  };
}

message CommitBookRequest {
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/Book"
    }];
  string tag = 2 [(google.api.field_behavior) = OPTIONAL];
}
```

- The method **must** use the `POST` HTTP verb.

> - The method **should** return a long running operation or resource revision.

Reviewer: the LRO would also return the resource revision, correct? if so, that should be clarified.

- The request message **must** include the `name` field.
- The resource ID of the resource revision **should** be a string and contain a
  short, automatically-generated random string. A good rule of thumb is the last
  eight characters of a [UUID version 4 (random)](UUID4).

> - `tag` field is optional. If provided, revision created will have the tag set.

Reviewer: since tag is mentioned here, I think the `tagging revisions` section should probably move up so the reader is familiar with the concept before arriving here.

### Getting a revision

Services `must` implement the standard `Get` method ([AIP-131][])

### Tagging revisions

There are two types of tags: server and user specified.

#### Server specified

Services **may** reserve specific tags to be server-specified (e.g. `current`,
`latest`), which are read-only and managed by the service.

> - If a `latest` tag exists, it **must** represents the most recent created

Reviewer: ```suggestion
- If a `latest` tag exists, it **must** represent the most recent created
```

Reviewer: most recently created revision or the most recent version of the resource? This terminology might make it a little ambiguous.

  revision.
 The content of `publishers/{publisher}/books/{book}/revisions/latest` and
  `publishers/{publisher}/books/{book}` can differ. Because the latest revision

>   may be different from the current state of the resource.

Reviewer 3: Does this mean it is up to the implementation to decide what a GET on the resource (as in no child revision specified) means?

So one could allow user specific revisions (such as the custom commit API example) which would not necessarily have its changes reflected on what is returned by actual resource?

Reviewer: a GET on the resource should return the resource in it's current state. That's true regardless of revisions. 

> So one could allow user specific revisions (such as the custom commit API example) which would not necessarily have its changes reflected on what is returned by actual resource?

Correct, that is the purpose of revisions: to store the state of a resource at a point in time. See "When resource revisions are useful" for an enumeration of use cases.

Did you have a different expectation from a revisions collection?

Reviewer 3: I think what I didn't quite understand is the fact that the documentation mentions that `latest` is not necessarily the same as `current`. If so, what is the definition of `current`? Is that left to the API to define?

Regarding the CustomCommit question, I think I just need to re-read the changes again. Almost feels like one could allow for requests where the user supplies the revision value through a custom method and if that's the case why have a concept of tags? Probably missed something somewhere.

```

> GET /v1/publishers/{publisher}/books/{book}/revisions/{reserved_keyword}

Reviewer: ```suggestion
GET /v1/publishers/{publisher}/books/{book}/revisions/{reserved_id}
```

There is no definition of "keyword": these are reserved resource IDs.

```

#### User specified

User specified tags are tags provided by the user when tagging specific resource
revision.

APIs implementing resource revisions **may** provide a mechanism for users to
tag a specific revision with a user provided name by implementing a "Tag
Revision" custom method:

```proto
rpc TagBookRevision(TagBookRevisionRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*/revisions/*}:tag"
    body: "*"
  };
}
```

```proto
message TagBookRevisionRequest {
  // The name of the book revision to be tagged.
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];

>   // The tag to apply. The tag should be at most 40 characters, and match

Reviewer: I think we spoke about this already, but tags are almost resource IDs as this point. This whole request could be reformatted in terms of a create.

  // `[a-z][a-z0-9-]{3,38}[a-z0-9]`.
  string tag = 2 [(google.api.field_behavior) = REQUIRED];
}
```
- The `name` field **should** require an explicit resource revision to be
  provided.
  - The field **should** be [annotated as required][aip-203].
  - The field **should** identify the [resource type][aip-123] that it
    references.
- The `tag` field **should** be [annotated as required][aip-203].
  - Additionally, tags **should** restrict letters to lower-case.
- Once a revision is tagged, the API **must** support using the tag in place of
  the revision ID in `name` fields.
  - If the user sends a tag or revision ID, the API **must** return the revision
    ID in the resource's `name` field.
- If the user calls the `Tag` method with an existing tag, the request **must**
  succeed and the tag updated to point to the new requested revision ID. This
  allows users to write code against specific tags (e.g. `published`) and the
  revision can change in the background with no code change.

> ### Listing revisions

Reviewer: maybe have this up by "Getting revisions"? It's just enumerating resource-oriented requirements.

APIs **may** store a revision history for a resource if it is useful to users.

APIs implementing resources with a revision history **should** abstract resource revisions as nested collection of the resource. The
resource revision type **must** inherit from the parent resource and has a field being the parent resource type.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;

   // The snapshot of the book
   Book book = 2 [(google.api.field_behavior) = IMMUTABLE];

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // tags set on the resource revisions by service or user.
  repeated string tags = 4;
  [(google.api.field_behavior) = OUTPUT_ONLY];
}
```

> - The resource revision **must** contain a field representing the snapshot of the parent resource.

Reviewer: ```suggestion
- The resource revision **must** contain a field representing the resource at the point in time the revision was created.
```

I worry about the term "snapshot" - so thinking this wording might be clearer?

> - The resource revision **must** contain a `create_time` field, which
>   **must** be a `google.protobuf.Timestamp` (see [AIP-142][]).

Reviewer: ```suggestion
- The resource revision **must** contain a `create_time` field (see [AIP-142][]).
```

I'm thinking this is better because we don't duplicate guidance on the type of create time - we have a few AIPs that are hard to keep in sync because of duplicated guidance.

Reviewer: this wasn't applied for some reason, please add.

> - The resource revision **may** contain a repeated field `tags` to represent tags that was set on the resource revision.

Reviewer 2: Last I checked, we were still unclear about what the exact use cases were for tags, and the best design. Might it be worth leaving tags out entirely to start with?

### When resource revisions are useful

> - When it is valuable to expose older versions of a resource via an API. This can avoid the overhead of the cusotmers having to write their own API to store and enable retrieval of revisions.

Reviewer: ```suggestion
- When it is valuable to expose older versions of a resource via an API. This can avoid the overhead of the customers having to write their own API to store and enable retrieval of revisions.
```

Also make sure we wrap @ 80 chars.

> - Oher resources depend on different revisions of a resource.

Reviewer: ```suggestion
- Other resources depend on different revisions of a resource.
```

- There is a need to visualize the change of a resource over time.

> ### Resource names for revisions

Reviewer: ```suggestion
### Resource names
```

"revisions" is implied by being a heading under this AIP.

> ### Resource names for revisions

Reviewer 2: I think we should move the "Committing revisions" part to here, as it's more conceptual - I'd also call it "creating revisions" rather than "committing" now that they're just resources.

Basically I'd personally - and @toumorokoshi may disagree here - like all the aspects of "what are revisions about and what you do with them" before the details of "what do the resource names and messages look like".

Reviewer: I agree that the unique attributes of revisions should be at the top, although not strongly enough to block the PR on it :)

IOW: it'd be great to make Jons suggested changes before merging this.

When referring to specific revision of a resource, APIs must use the following syntax
`{resource_name}/revisions/{revision_id}`. For example:

publishers/123/books/les-miserables/revisions/c7cfa2a8

### Getting a revision

> Services `should` implement the standard `Get` method ([AIP-131][])

Reviewer 2: I'd suggest "must" here.

### Tagging revisions

There are two types of tags: server defined and user specified.

#### Server defined

Services **may** reserve specific tags to be server-defined (e.g. `current`, `latest`), which are read-only and managed by the service.

> - `latest` represents the most recent created revision. The content of `publishers/{publisher}/books/{book}/revisions/latest` and `publishers/{publisher}/books/{book}` can differ. Because the latest revision may be different from the current state of the resource.

Reviewer: ```suggestion
- If a `latest` tag exists, it **must** represents the most recent created revision. The content of `publishers/{publisher}/books/{book}/revisions/latest` and `publishers/{publisher}/books/{book}` can differ. Because the latest revision may be different from the current state of the resource.
```

Fix for AIP styling. also watch line length (80 chars).

```
GET /v1/publishers/{publisher}/books/{book}/revisions/{reserved_keyword}
```

> #### User Specified

Reviewer: ```suggestion
#### Tagging revisions
```

There's not much a reason to rename this section - and renaming sections will break links. So keep it the same?

At minimum I would update to `User Specified Tags`.

    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];

  // The tag to apply.
  // The tag should be at most 40 characters, and match `[a-z][a-z0-9-]{3,38}[a-z0-9]`.
  string tag = 2 [(google.api.field_behavior) = REQUIRED];
}
```
- The `name` field **should** require an explicit resource revision to be provided.
  - The field **should** be [annotated as required][aip-203].
  - The field **should** identify the [resource type][aip-123] that it
    references.
- The `tag` field **should** be [annotated as required][aip-203].
  - Additionally, tags **should** restrict letters to lower-case.
- Once a revision is tagged, the API **must** support using the tag in place of
  the revision ID in `name` fields.
  - If the user sends a tag or revision ID, the API **must** return the revision ID in the resource's
    `name` field.
- If the user calls the `Tag` method with an existing tag, the request **must**
  succeed and the tag updated to point to the new requested revision ID. This
  allows users to write code against specific tags (e.g. `published`) and the
  revision can change in the background with no code change.
### Listing revisions

> Services `should` implement the standard `List` method ([AIP-132][])

Reviewer: this is a *must* according to AIP-121.

  succeed and the tag updated to point to the new requested revision ID. This
  allows users to write code against specific tags (e.g. `published`) and the
  revision can change in the background with no code change.
### Listing revisions

APIs implementing resource revisions `should` provide a standard List method for listing the revision history for a resource

> ```proto

Reviewer: proto is unnescessary, I'd like just link to AIP-132 and state that it has to be compatible with AIP-132 get.

Author: removed and updated

If a child resource is a child of a single revision, the child resource's name
**must** always explicitly include the parent's revision ID:

>     publishers/123/books/les-miserables/revisions/c7cfa2a8/pages/42

Reviewer 2: I think we need more detail about the messages for child resources. In the previous scheme, we didn't have different messages for revisions and non-revisions, because they used the same resource name. Now we've got different messages for the top-level resources - do we have that for child resources as well? If not, is there any way of knowing which top-level revision a child resource was fetched from?

If necessary, APIs **may** explicitly support listing child resources across
parent revisions by accepting the `-` syntax. For example:

>     GET /v1/publishers/123/books/les-miserables/revisions/-/pages

Reviewer: Isn't this guidance redundant? the former is obvious: where do you put a child of a revision?

the latter is basically AIP-159. 

I feel like this part could just be removed.

- Commit a new revision when something important happens
- Commit a new revision when the user specifically asks

APIs **may** use any of these strategies. APIs that want to commit a revision on user request should
handle this with a custom method on the resource:

```proto

> rpc CommitBookRevision(CommitBookRevisionRequest) returns (BookRevision) {

Reviewer 2: Should this still be Commit, or should it just be Create? (It's just creating a resource - but there are limits in terms of what can be specified of course.)

message ListBookRevisionsResponse {
  // The revisions of the book.
  repeated BookRevision books = 1;

  // A token that can be sent as `page_token` to retrieve the next page.
  // If this field is omitted, there are no subsequent pages.
  string next_page_token = 2;
}
```

> - Revisions must be ordered in reverse chronological order. An `order_by` field **should not** be provided in the list request.

Reviewer: is this necessary? I'm not clear on why one couldn't order revisions inf they wanted to.

Author: I think this is a useful default behavior. Updated the sentence. WDYT?

Reviewer: We definitely should have an `order_by` field - not having it produces exceptions in things like the API linter and also force tooling to handle partial List endpoints.

I'm fine with keeping the guidance on reverse chronological order for now (and btw makes sure to use `**must**` to emphasize it's and RFC-2119 term.

method:

```proto

> rpc RollbackBook(RollbackBookRequest) returns (BookRevision) {

Reviewer 2: I'm not *sure* whether "rollback" is the right terminology to use now that the resources are separate.
Revert and restore seem just as suitable... hmm. (This may have been discussed internally already.)

Reviewer: +1 to restore. I think semantically this would make sense to, if there's data-plane information that's attached to a collection as well.

- Commit a new revision when something important happens
- Commit a new revision when the user specifically asks

APIs **may** use any of these strategies. APIs that want to commit a revision on user request should
handle this with a custom method on the resource:

```proto

> rpc CommitBookRevision(CommitBookRevisionRequest) returns (BookRevision) {

Reviewer: should a tag field be in the request as well? it might make it easier to tag a specific revision as you're creating it.

```

- The method **must** use the `POST` HTTP verb.

> - The method **should** return a long running operation or resource revision.

Reviewer: similar to above, the LRO would return the revision, correct?

Revisions are sometimes expensive to store, and there are valid use cases to
want to remove one or more revisions from a resource's revision history.

APIs **may** define a standard `Delete` ([AIP-135][]) method to delete resource revisions

> - Resource revisions **should** be automatically deleted on parent resource deletion.

Reviewer 2: What about soft deletion?

Author: I think for soft delete, all the revision resources should be soft deleted as well. Updated the paragraph.

> - If resource revisions meant to have longer lifespan than the parent resource, resource and resource revision

Reviewer 2: ```suggestion
- If resource revisions are meant to have longer lifespan than the parent resource, resource and resource revision
```

(Probably make this change yourself - GitHub suggestions from me cause CLA-bot issues.)

> should be side by side resource collections.

Reviewer 2: I don't know what this means. More detail definitely required.

Author: Move it to the top of the page where "abstract revisions as nested collections are not proper"

## Rationale

### Deprecate the @-based approach

Historically, Resource revision was recommended to be referenced by `{resource_id}@{revision_id}`.

It has several disadvantages:

-  List revisions is a custom method (:listRevisions) on the resource collection
-  Delete revision is a custom method on the resource collection
-  Not visible in api discovery doc

> -  Resource ID cannot use `@`

Reviewer 2: It might be worth listing the benefits that we're losing as well though. In particular, the ability to handle a resource revision or a normal resource in the same way, both in terms of code and resource names.

I'm pretty sure we'll want to add the ability to specify that a resource name is "either X or Y" where the main use case is "either a resource or a resource revision".

Reviewer 2: It's possibly also worth noting that although this makes revisions feel like "normal" resources in some ways, there are still significant differences between them and regular ones - e.g. creation patterns.

Reviewer: FWIW I think the creation could also be normalized now - a tag on a revision is a skip and a hop away from a resource ID.

I think it's fair to include the advantages, as well as explain how they might achieved in the new world. For now I think we're talking about a "*" resource reference since we don't have the semantics to express references to more than one resource.

Revisions are sometimes expensive to store, and there are valid use cases to
want to remove one or more revisions from a resource's revision history.

> APIs **may** define a standard `Delete` ([AIP-135][]) method to delete resource

Reviewer: this feels a little unnecessary to me - this, along with Get / List could be probably consolidated to a section like "A revision is a collection" or something along those lines, then enumerate that Get / List are required and delete is optional.

Reviewer: also, similar to "Getting Revisions" or "Listing Revision", this should be it's own section, correct?

revisions

> - Resource revisions **should** be automatically deleted on parent resource

Reviewer: I'd remove this. there's already discussion around cascading deletion: https://google.aip.dev/135#cascading-delete. revisions should follow those rules.

Revisions are sometimes expensive to store, and there are valid use cases to
want to remove one or more revisions from a resource's revision history.

> APIs **may** define a standard `Delete` ([AIP-135][]) method to delete resource revisions

Reviewer: This should be in a section that talks about revisions being a child collection and following the standard method AIPs (131,132,133,134,135).

  // The name of the book revision to be deleted, with a revision ID explicitly
  // included.
  //
  // Example: publishers/123/books/les-miserables/revisions/c7cfa2a8
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];
}
```

- Resource revisions **should** be automatically deleted on parent resource deletion.
- If resource revisions meant to have longer lifespan than the parent resource, resource and resource revision
should be side by side resource collections.

> ## Changelog

Reviewer: Please include a rationale and history section to explain the change. Example: https://google.aip.dev/203#rationale.

I think I'd primarily talk about how the goal is to make revisions more of a sub-resource than an extension of an existing resource.
